### ISSUE TITLE
call_remote!() with same agent key to a different zome fails when the callee function tries to fetch an entry committed by another agent through get_links! and get!

### ISSUE THREAD

**Issue by tatssato** [holochain/holochain#397]
I have two zomes in one cell, `zomeone` and `zometwo`. 
`zomeone` has `create_foo` and `get_author_of_foo` zome fn.
`zometwo` has `get_author_of_foo_from_zome_one` which invokes `call_remote!()` to the same agent but to `zomeone` to call `get_author_of_foo`.

```rust
// zomeone

use hdk3::prelude::*;

#[hdk_entry(id = "foo", visibility = "public" )]
#[derive(Clone)]
struct Foo (String);

#[derive(Deserialize, Serialize, SerializedBytes)]
struct StringWrapper(String);

entry_defs![
    Path::entry_def(),
    Foo::entry_def()
];

#[hdk_extern]
fn create_foo(_: ()) -> ExternResult<Foo> {
    let foo = Foo("test".to_owned());
    create_entry!(foo.clone())?;
    let path = Path::from("bar".to_owned());
    path.ensure()?;
    create_link!(
        hash_entry!(path)?,
        hash_entry!(foo.clone())?,
        LinkTag::new("testing".to_owned())
    )?;
    Ok(foo)
}

#[hdk_extern]
fn get_author_of_foo(input: StringWrapper) -> ExternResult<AgentPubKey> {
    let path = Path::from(input.0);
    let path_hash = hash_entry!(path.clone())?;
    let links = get_links!(hash_entry!(path)?)?;
    debug!("hash of the path is {:#?}", path_hash)?;
    debug!("links are {:#?}", links)?;
    let link = links.into_inner()[0].clone();
    match get!(link.target)? {
        Some(element) => {
            let header_details = element.header();
            Ok(header_details.author().to_owned())
        },
        _ => Err(HdkError::Wasm(WasmError::Zome(String::from(
            "element not found"
        ))))
    }
}
```

```rust
// zometwo

use hdk3::prelude::*;

#[derive(Deserialize, Serialize, SerializedBytes)]
struct StringWrapper(String);

#[hdk_extern]
fn get_author_of_foo_from_zomeone(_: ()) -> ExternResult<AgentPubKey> {
    let my_agent_pubkey = agent_info!()?.agent_latest_pubkey;
    let function_name = zome::FunctionName("get_author_of_foo".to_owned());
    let payload: SerializedBytes = StringWrapper("bar".to_owned()).try_into()?;

    match call_remote!(my_agent_pubkey, "zomeone".into(), function_name, None, payload)? {
        ZomeCallResponse::Ok(output) => {
            let sb = output.into_inner();
            let agent_pubkey: AgentPubKey = sb.try_into()?;
            Ok(agent_pubkey)
        },
        ZomeCallResponse::Unauthorized => {
            Err(HdkError::Wasm(WasmError::Zome(
                "this agent has no proper authorization".to_owned()
            )))
        },
    }
}
```

I have four tryorama test scenarios for this.
The following scenarios pass the test.

- Scenario 1: **alice** calls `create_foo` and then calls `get_author_of_foo` **directly from `zomeone`.** `get_author_of_foo` returns alice's key (since she authored `foo`) so no problem.

- Scenario 2: **bobby** calls `create_foo` and then **alice** calls `get_author_of_foo` **directly from `zomeone`**. `get_author_of_foo` returns bobby's key so no problem.

- Scenario 3: **alice** calls `create_foo` from `zomeone` and then calls `get_author_of_foo_from_zomeone` **from `zometwo`.** `get_author_of_foo_from_zomeone` returns alice's key (since she authored `foo`) so no problem here as well.

Here's where things gets a little bit weird.
- Scenario 4: **bobby** calls `create_foo` from `zomeone` and then **alice** calls `get_author_of_foo_from_zomeone` **from `zometwo`.** `get_author_of_foo_from_zomeone` **should return bobby's key but doesn't.** Instead, the tryorama fails with the following error.

```
{
  type: 'Error',
  data: {
    type: 'RibosomeError',
    data: 'Wasm error while working with Ribosome: CallError("RuntimeError: unknown error")'
  }
}
```
doing debug!() on `get_author_of_foo` shows that path hash is being generated properly from the `input.0`
However, `links` are empty only in Scenario 4. The difference here is that _alice is trying to get bobby's entry with `get!(link.target)` in a call_remote fashion._

There is also this Debug output from conductor(being outputted a couple of times before the test just fails. probably trying to get the dependency but then after a lot of tries, it can't get it and just simply stops trying).

```
☮ Oct 05 19:45:19.326 DEBUG app_validation_workflow: holochain::core::workflow::app_validation_workflow: ops=[ValidationLimboValue { status: SysValidated, pending_dependencies: PendingDependencies { pending: [] }, op: StoreEntry(HeaderHash(uhCkkIVWv9PGrTexDs-yiFmIrTxhfz_WUD4G9ZJ5UZkns5CW-xZy0), EntryHash(uhCEkzcdzyczLwOd8cNE94Bxdv9IoS4NBoBn-rpXl6F5yqDWpQPkn), AnyDhtHash(uhCEkzcdzyczLwOd8cNE94Bxdv9IoS4NBoBn-rpXl6F5yqDWpQPkn)), basis: AnyDhtHash(uhCEkzcdzyczLwOd8cNE94Bxdv9IoS4NBoBn-rpXl6F5yqDWpQPkn), time_added: Timestamp(1601894719, 316067100), last_try: Some(Timestamp(1601894719, 326096900)), num_tries: 1 }] awaiting_ops=[ValidationLimboValue { status: PendingValidation, pending_dependencies: PendingDependencies { pending: [FixedElement(DhtOpHash(uhCQk18c3Ocp8gLn_KNebDyLfuCXOVhbO4nyDNg683CpAI3zANiNq))] }, op: RegisterAgentActivity(HeaderHash(uhCkkxHqy2WHnkgtrD_2ulQkyQJwf_L5WkoyCgdrsNkXWJGIowwHJ), AnyDhtHash(uhCEke1j8Z2a-_min0h0pGuEMcYlo_V1l1mt9OtBuywKmHlg4L_R-)), basis: AnyDhtHash(uhCEke1j8Z2a-_min0h0pGuEMcYlo_V1l1mt9OtBuywKmHlg4L_R-), time_added: Timestamp(1601894719, 317450400), last_try: Some(Timestamp(1601894719, 322516200)), num_tries: 1 }, ValidationLimboValue { status: PendingValidation, pending_dependencies: PendingDependencies { pending: [AnyElement(DhtOpHash(uhCQk5L33gc-JG_nmzedLV-Y0stz8AjULMUomjDYcYl_-wTID99Vj))] }, op: RegisterAddLink(HeaderHash(uhCkkxHqy2WHnkgtrD_2ulQkyQJwf_L5WkoyCgdrsNkXWJGIowwHJ), AnyDhtHash(uhCEkVYGAv-lxOIpyeiKKzL56ZFA6Co_5_S8hraN8wRQyO2NmLur8)), basis: AnyDhtHash(uhCEkVYGAv-lxOIpyeiKKzL56ZFA6Co_5_S8hraN8wRQyO2NmLur8), time_added: Timestamp(1601894719, 314977400), last_try: Some(Timestamp(1601894719, 319342000)), num_tries: 2 }]
``` 

```typescript
import { Orchestrator } from '@holochain/tryorama';
import { Config } from '@holochain/tryorama';

const orchestrator = new Orchestrator();

const config = Config.gen({
    alice: Config.dna('../testlab.dna.gz', null),
    bobby: Config.dna('../testlab.dna.gz', null),
  })

const delay = (ms) => new Promise((r) => setTimeout(r, ms))

orchestrator.registerScenario('Scenario 1: directly calling create and get from zomeone to get own key from entry', async (s, t) => {
  // spawn the conductor process
  const { conductor } = await s.players({ conductor: config });
  await conductor.spawn();

  const [dna_hash_1, agent_pubkey_alice] = conductor.cellId('alice');

  await conductor.call("alice", "zomeone", "create_foo", null);
  await delay(1000);

  // Works with no problem
  let get_alice_key = await conductor.call("alice", "zomeone", "get_author_of_foo", "bar");

  t.deepEqual(get_alice_key, agent_pubkey_alice);
});

orchestrator.registerScenario('Scenario 2: directly calling create and get from zomeone to get bobby key from entry', async (s, t) => {
  // spawn the conductor process
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  const [dna_hash_2, agent_pubkey_bobby] = conductor.cellId('bobby');

  await conductor.call("bobby", "zomeone", "create_foo", null);
  await delay(1000);

  // works with no problem too
  let get_bobby_key = await conductor.call("alice", "zomeone", "get_author_of_foo", "bar");

  t.deepEqual(get_bobby_key, agent_pubkey_bobby);
});

orchestrator.registerScenario('Scenario 3: directly calling create but remotely call get zometwo to get own key from entry', async (s, t) => {
  // spawn the conductor process
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  const [dna_hash_1, agent_pubkey_alice] = conductor.cellId('alice');

  await conductor.call("alice", "zomeone", "create_foo", null);
  await delay(1000);

  // this works with no problem as well
  let get_alice_key = await conductor.call("alice", "zometwo", "get_author_of_foo_from_zomeone", null);

  t.deepEqual(get_alice_key, agent_pubkey_alice);

});

orchestrator.registerScenario('Scenario 4: directly calling create but remotely call get zometwo to get both key from entry', async (s, t) => {
  // spawn the conductor process
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  const [dna_hash_2, agent_pubkey_bobby] = conductor.cellId('bobby');

  await conductor.call("bobby", "zomeone", "create_foo", null);
  await delay(1000);

  // now this doesn't work because the links return an empty array inside the get_author function
  let get_bobby_key = await conductor.call("alice", "zometwo", "get_author_of_foo_from_zomeone", null);

  t.deepEqual(get_bobby_key, agent_pubkey_bobby);
});


orchestrator.run()
```

Any help would be appreciated.
Thank you.

**Comment by tatssato** [comment]
updated my holochain to commit `4714fbddc48f1f9987c9f574b8d06510ea58ab16`

did the test again and now it shows this `WARN`

```
★ Oct 05 22:33:56.725  WARN sys_validation_workflow: holochain::core::workflow::sys_validation_workflow: agent=alice msg="DhtOp has failed system validation" op=StoreEntry(Signature(0xf4d1cde09b16b72c7abf27b710f7172441b2e703981cbd7a190491be5da0aec39852b68f48a593c5c0c1fc0b0d543da82aa64843c1c68f6062ab30d6b6937a00), Create(Create { author: AgentPubKey(uhCAke1j8Z2a-_min0h0pGuEMcYlo_V1l1mt9OtBuywKmHlg4L_R-), timestamp: Timestamp(1601904836, 296207800), header_seq: 5, prev_header: HeaderHash(uhCkkztWwr1NNr-hDn0UNLrPo2f1pZPuDxejWt_6FGtOrKv-MIhOd), entry_type: App(AppEntryType { id: EntryDefIndex(0), zome_id: ZomeId(0), visibility: Public }), entry_hash: EntryHash(uhCEkVYGAv-lxOIpyeiKKzL56ZFA6Co_5_S8hraN8wRQyO2NmLur8) }), App(AppEntryBytes([[98,0,0,0,97,0,0,0,114,0,0,0]]))) error=DepMissingFromDht(AnyDhtHash(uhCkkztWwr1NNr-hDn0UNLrPo2f1pZPuDxejWt_6FGtOrKv-MIhOd)) error_msg=The dependency AnyDhtHash(uhCkkztWwr1NNr-hDn0UNLrPo2f1pZPuDxejWt_6FGtOrKv-MIhOd) was not found on the DHT

```

**Comment by freesig** [comment]
Having a look into this now. I will try and recreate your test as a rust test and see what happens.
The warning is expected as it just means at the time of validating the entry the previous header had not yet been received. This can happen when something gets to the incoming validation slightly ahead of its dependencies

**Comment by freesig** [comment]
Thanks for such a details issue.
I've made a more simplified rust test at #402 but it seems to be all working fine so it's not the remote call. Perhaps there's something up with how you are constructing the paths?
Let me know if there's more I can add to that test but for now I might leave this unless we can find an issue.

**Comment by freesig** [comment]
I think the error your getting is from `let link = links.into_inner()[0].clone();`
It's generally not a good idea to use indices in a wasm and prefer `.get(0)` as indices panic when they are out of range.
It could just be that the delay isn't enough time for the links to get to alice but 1 second should be plenty. You could try setting it to 10 seconds just to see but it makes me think something isn't getting through validation.

**Comment by tatssato** [comment]
Thanks for responding so quickly!
Thanks also for pointing out the use of indices. I took that whole thing out and the error I was getting is gone.
I've simplified the test as well to match yours here #402 and just get the links and return it.
I tested both cases.
- scenario 1: base is a `Path`
- scenario 2: base is an entry `Foo("foo")`
```rust 
//ZOMEONE
#[hdk_extern]
fn create_foo_and_link_to_path(_: ()) -> ExternResult<Foo> {
    let foo = Foo("foo".to_owned());
    create_entry!(&foo)?;
    let path = Path::from("bar".to_owned());
    path.ensure()?;
    create_link!(
        hash_entry!(&path)?,
        hash_entry!(&foo)?,
        LinkTag::new("foo->path".to_owned())
    )?;
    Ok(foo)
}

#[hdk_extern]
fn get_links_from_path(input: StringWrapper) -> ExternResult<Links> {
    let path = Path::from(input.0);
    let path_hash = hash_entry!(&path)?;
    let links = get_links!(path_hash.clone())?;
    Ok(links)
}

#[hdk_extern]
fn create_and_link_foo(_: ()) -> ExternResult<()> {
    let base = Foo("foo".to_string());
    let target = Foo("foofoo".to_string());
    let base_hash = hash_entry!(&base)?;
    let target_hash = hash_entry!(&base)?;
    let tag = LinkTag::new("foos");
    create_entry!(&base)?;
    create_entry!(&target)?;
    create_link!(base_hash, target_hash, tag)?;
    Ok(())
}

#[hdk_extern]
fn get_links_from_foo(input: StringWrapper) -> ExternResult<Links> {
    let base = Foo(input.0.to_owned());
    let base_hash = hash_entry!(&base)?;
    let links = get_links!(base_hash)?;
    Ok(links)
}

//ZOMETWO
#[hdk_extern]
fn get_links_from_path_from_zomeone(_: ()) -> ExternResult<Links> {
    let my_agent_pubkey = agent_info!()?.agent_latest_pubkey;
    let function_name = zome::FunctionName("get_links_from_path".to_owned());
    let payload: SerializedBytes = StringWrapper("bar".to_owned()).try_into()?;

    match call_remote!(my_agent_pubkey, "zomeone".into(), function_name, None, payload)? {
        ZomeCallResponse::Ok(output) => {
            let sb = output.into_inner();
            let links: Links = sb.try_into()?;
            Ok(links)
        },
        ZomeCallResponse::Unauthorized => {
            Err(HdkError::Wasm(WasmError::Zome(
                "this agent has no proper authorization".to_owned()
            )))
        },
    }
}

#[hdk_extern]
fn get_links_from_foo_from_zomeone(_: ()) -> ExternResult<Links> {
    let my_agent_pubkey = agent_info!()?.agent_latest_pubkey;
    let function_name = zome::FunctionName("get_links_from_foo".to_owned());
    let payload: SerializedBytes = StringWrapper("foo".to_owned()).try_into()?;

    match call_remote!(my_agent_pubkey, "zomeone".into(), function_name, None, payload)? {
        ZomeCallResponse::Ok(output) => {
            let sb = output.into_inner();
            let links: Links = sb.try_into()?;
            Ok(links)
        },
        ZomeCallResponse::Unauthorized => {
            Err(HdkError::Wasm(WasmError::Zome(
                "this agent has no proper authorization".to_owned()
            )))
        },
    }
}
```

Tryorama test is as follows:

```Typescript
orchestrator.registerScenario('Scenario 4: bobby create and alice get links from path remotely from zometwo', async (s, t) => {
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  await conductor.call("bobby", "zomeone", "create_foo_and_link_to_path", null);
  await delay(100000);

  let links = await conductor.call("alice", "zometwo", "get_links_from_path_from_zomeone", null);

  t.deepEqual(links.length, 1);
});

orchestrator.registerScenario('Scenario 5: bobby creates and alice get links from foo from zomeone', async (s, t) => {
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  await conductor.call("bobby", "zomeone", "create_and_link_foo", null);
  await delay(100000);

  let links = await conductor.call("alice", "zometwo", "get_links_from_foo_from_zomeone", null);

  t.deepEqual(links.length, 1);
});
```

Even with a delay of 1min30sec, both tests fail as `links.length` is 0.
I tried these scenario for `create_and_link_foo` and `get_links_from_foo` and all of them returned 1 link which made the test pass. 
- alice create and alice get links directly from zomeone -> return 1 link: test passes
- bobby create and alice get links directly from zomeone -> return 1 link: test passes
- alice create and alice get links remotely from zometwo -> return 1 link: test passes

So the only time the `links.length` is 0 is when _bobby create and alice get links remotely from zometwo_ still.
But since this test passsed in your rust test, the only difference is Im using Tryorama to test it and maybe _that_ is the problem...

**Comment by tatssato** [comment]
FWIW, I also tried this scenario and links still returned empty in tryorama.
TL;DR
1. bobbo grants capability for function `get_links_from_foo` to be called unrestrictedly.
2. alice calls `create_and_link_foo` to commit `Foo("foo")` and `Foo("foofoo")`, and create a link (base: `Foo("foo")`, target: `Foo("foofoo")`.
3. alice remotely calls into bobbo's cell to invoke `get_links_from_foo` to fetch the link to `Foo("foofoo")` from `Foo("foo")` (here, even if alice committed the entry, she is calling `get_links_from_foo` from bobbo's cell).

From this, it seems like using `call_remote` to fetch links from a cell where that same cell did not commit and linked the entry is the problem. Again, if I dont use call_remote, alice can fetch with no problem the link to an entry bobbo committed and linked (which is a case where cell1 is getting links to an entry cell2 committed).

Here is the code for the scenario above
```rust
#[hdk_extern]
fn create_cap_grant(_: ()) -> ExternResult<()> {
    let mut functions: GrantedFunctions = HashSet::new();
    functions.insert((zome_info!()?.zome_name, "get_links_from_foo".into()));
    create_cap_grant!(
        CapGrantEntry {
            tag: "".into(),
            // empty access converts to unrestricted
            access: ().into(),
            functions,
        }
    )?;
    Ok(())
}

#[hdk_extern]
fn create_and_link_foo(_: ()) -> ExternResult<()> {
    let base = Foo("foo".to_string());
    let target = Foo("foofoo".to_string());
    let base_hash = hash_entry!(&base)?;
    let target_hash = hash_entry!(&base)?;
    let tag = LinkTag::new("foos");
    create_entry!(&base)?;
    create_entry!(&target)?;
    create_link!(base_hash, target_hash, tag)?;
    Ok(())
}

#[hdk_extern]
fn get_links_from_foo(_: ()) -> ExternResult<Links> {
    let base = Foo("foo".to_owned());
    let base_hash = hash_entry!(&base)?;
    let links = get_links!(base_hash)?;
    Ok(links)
}

#[hdk_extern]
fn caller(agent: AgentPubKey) -> ExternResult<Links> {
    let function_name = zome::FunctionName("get_links_from_foo".to_owned());

    match call_remote!(
        agent, 
        "zomeone".into(),
        function_name, 
        None,
        ().try_into()?
    )? {
        ZomeCallResponse::Ok(output) => {
            let sb = output.into_inner();
            let links: Links = sb.try_into()?;
            Ok(links)
        },
        ZomeCallResponse::Unauthorized => {
            Err(HdkError::Wasm(WasmError::Zome(
                "this agent has no proper authorization".to_owned()
            )))
        },
    }
}
``` 

Here's the tryorama test where `links.length` is 0 when it should be 1.
```typescript
orchestrator.registerScenario('Scenario 6: alice creates and alice calls remotely to bobbos cell to get links ', async (s, t) => {
  const { conductor } = await s.players({ conductor: config })
  await conductor.spawn()

  const [dna_hash, agent_pubkey_bobbo] = conductor.cellId('bobbo');
  // bobbo grants capability for function get_links_from_foo to be called unrestrictedly.
  await conductor.call("bobbo", "zomeone", "create_cap_grant", null);

  // alice commit Foo("foo") and Foo("foofoo"), and create a link (base: Foo("foo"), target: Foo("foofoo").
  await conductor.call("alice", "zomeone", "create_and_link_foo", null);
  await delay(10000);

  // alice calls function `caller` to `call_remote`, which invokes get_links_from_foo in bobbo's cell
  let links = await conductor.call("alice", "zomeone", "caller", agent_pubkey_bobbo);
  console.log("get links remotely");
  console.log(links);

  // this test fails as links returned from alice's call_remote to bobbo's to be 0
  t.deepEqual(links.length, 1);
});
```
Hope this helps.

**Comment by freesig** [comment]
Hey I think it's just the agent key for the call_remote was set to the wrong agent. See #402  for the changes. If that's the fix we can close this issue.

**Comment by tatssato** [comment]
hi! Mhmm but we're testing a call_remote to the same agent (same cell) but different zome right? In that case, the agent key you pass in should just be the same as the agent key of the caller since bobbo is doing call_remote to himself... and I tried passing the agent key from tryorama and I still get an empty result :( Also, I'm starting to think that #402 this also only passed the test because bobbo first got the link by calling `get_links_on_foo` directly (which means the link was cached) then when he invoked call_remote, the link was already in the cache which was returned to him successfully... I added a comment here suggesting to [rerun](https://github.com/holochain/holochain/pull/402/files#r506050421) the test

**Comment by tatssato** [comment]
This issue was fixed by using `call` (which was available in `develop` starting [here](https://github.com/holochain/holochain/commit/e8b3efca5d2e8be58d4046355265eb4de58bff33) instead of `call_remote`). 
will close it since this issue is fixed for us,.
Thank you for all the help we got from the team, especially @freesig!
